### PR TITLE
CERD-210 > graph

### PR DESCRIPTION
**Pull request by chrisunl**
WIP

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,7 +13,14 @@
...
     "devDependencies": {
+<<<<<<< HEAD
+        "lerna": "latest"
+    },
+    "dependencies": {
+        "@types/d3": "^5.0.0"
+=======
         "lerna": "latest",
         "puppeteer": "^1.8.0"
+>>>>>>> master
```
Comment (drochgenius) [inline package.json]: Please make sure you resolve conflicts before pushing them in your branch

File: packages/plot-graph/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Assessment System Text Input Component
+
+## Usage
+
+```html
+<!-- basic usage -->
+    <plot-graph xValues='0,100' equation='y=x'>
+        <response-validation slot="feedback" expected="y=x" score="4" strategy="fuzzyMatch">
```
Comment (drochgenius) [inline packages/plot-graph/README.md]: Does feedback apply to your interaction?
Isn't this purely exploratory?

File: packages/plot-graph/src/components/plot-graph.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
+import { ComponentBase, html, TemplateResult, Feedback, FeedbackMessage, Strategy, applyMixins } from "@hmh/component-base/dist/index";
+import { MDCTextField } from "@material/textfield/index";
+import { ResponseValidation } from '@hmh/component-base/dist/index';
+import * as d3 from "d3";
+
+class Line {
+    private static _nullValue: number = -10000;
```
Comment (drochgenius) [inline packages/plot-graph/src/components/plot-graph.ts]: ?

File: packages/plot-graph/src/components/plot-graph.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    private _startX: number = Line._nullValue;
+    private _endX: number = Line._nullValue;
+    private _startY: number = Line._nullValue;
+    private _endY: number = Line._nullValue;
```
Comment (drochgenius) [inline packages/plot-graph/src/components/plot-graph.ts]: Not sure why you're initializing all numeric values to `-1000`.
Do you knwo the JS number `Infinity`?

File: packages/plot-graph/src/components/plot-graph.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        return 'X: ' + this.startX.toString() + ' to ' + this.endX.toString() + 
+        ' Y: ' + this.startY.toString() + ' to ' + this.endY.toString();
+    }
+}
+
+/**
+ * `<plot-graph>`
+ * @demo ./demo/index.html
+ */
+export class PlotGraph extends ComponentBase<string> implements Feedback {
```
Comment (drochgenius) [inline packages/plot-graph/src/components/plot-graph.ts]: Again, not sure if we need to implement Feedback for this interaction

File: packages/plot-graph/src/components/plot-graph.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        `;
+    }
+
+    /**
+     * Get feedback classes for UI depending upon feedbackMessage.type
+     *
+     * @param  {FeedbackMessage} feedbackMessage
+     * @returns String
+     */
+    private _getFeedbackClass(feedbackMessage: FeedbackMessage, isContainer: boolean): string {
```
Comment (drochgenius) [inline packages/plot-graph/src/components/plot-graph.ts]: Do we need this?

File: packages/plot-graph/README.md (hunk 2 above)
Comment (chrisunl) [reply packages/plot-graph/README.md]: My understanding is that the user has to solve for Y if they know X?

File: packages/plot-graph/src/components/plot-graph.ts (hunk 3 above)
Comment (chrisunl) [reply packages/plot-graph/src/components/plot-graph.ts]: ?

File: packages/plot-graph/src/components/plot-graph.ts (hunk 4 above)
Comment (chrisunl) [reply packages/plot-graph/src/components/plot-graph.ts]: whoops. no, I didn't know that

File: packages/plot-graph/src/components/plot-graph.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (drochgenius) [reply packages/plot-graph/src/components/plot-graph.ts]: Hi @Dom-HMH-MTL , makes sense to me.
Please review the story on JIRA here, I've added the notes from today's discussion.
https://jira.hmhco.com/browse/CERD-210

Feel free to add/modify the description.
Let's catch up tomorrow on a white board

File: packages/plot-graph/package.json [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        "_generateDocs": "polymer analyze ./dist/components/plot-graph.js > ./dist/components/analysis.json"
+    },
+    "keywords": [],
+    "author": "Christopher Morris <christopher.morris@hmhco.com>",
+    "license": "ISC",
+    "dependencies": {
+        "@hmh/component-base": "latest",
+        "@material/button": "latest",
+        "@material/textfield": "latest",
+        "babel": "^6.23.0",
```
[the remark this answers is not in the document]
Comment (chrisunl) [reply packages/plot-graph/package.json]: Babel was needed for d3, so I guess it's a dev dependency. Not sure if I need nyc here, it just kind of got added.

[the remark this answers is not in the document]
Comment (drochgenius) [reply packages/plot-graph/package.json]: Babel for D3, we're not targeting ES5, and in any case we have Typescript, so it does not make any sense to me?
`nyc` you don't need.

File: packages/plot-graph/src/components/plot-graph.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                const prepareValue = this.prepareValue;
+                const dataset = d3.range(numberPoints).map(function(x: any) { return {"y": prepareValue(equation, x)}});
+
+                this.svgContainer.attr("width", this.graphSize)
+                .attr("height", this.graphSize)
+                .append("g")
+                .attr("transform", "translate(0,0)");
+
+                // 3. Call the x axis in a group tag
+                this.svgContainer.append("g")
```
Comment (drochgenius) [inline packages/plot-graph/src/components/plot-graph.ts]: Not sure if we want to add the axis for every equation.
My understanding is that all equations would share the same axis system.

File: packages/plot-graph/README.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            <span>Wooot!</span>
+        </response-validation>
+
+        <response-validation slot="feedback" score="0" feedback="negative" strategy="exactMatch">
+            <span>Try again</span>
+            <span>Oups, still not good</span>
+            <span>Too bad!</span>
+        </response-validation>
+    </plot-graph>
+```
```
Comment (drochgenius) [inline packages/plot-graph/README.md]: Please review the README @chrisunl

File: packages/plot-graph/src/components/plot-graph.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+    private graphSize: number = 500;
+    private svgContainer: any = null;
+    private renderedGraph: boolean = false;
+    
+    protected render(): TemplateResult {
+        return html`
+        <link rel="stylesheet" type="text/css" href="/dist/css/plot-graph.css">
+            <div id="canvas"></div>
+        <slot hidden name="options" class="options" @slotchange=${(evt: Event) => this._onSlotChanged(evt)}> </slot>
```
Comment (drochgenius) [inline packages/plot-graph/src/components/plot-graph.ts]: fix indent for readability

File: packages/plot-graph/src/unit/test-helpers.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import { PlotGraph } from '../components/plot-graph';
+
+const expect: any = chai.expect;
+
+export function checkComponentDOM(el: PlotGraph, params: { [key: string]: string | number | boolean } = {}): void {
+    const shadowRoot: ShadowRoot = el.shadowRoot;
+    expect(shadowRoot).not.to.be.null;
+}
```
Comment (drochgenius) [inline packages/plot-graph/src/unit/test-helpers.ts]: The implementation looks good.
Please cover with unit tests.

**Comment by drochgenius** [comment]
Hey @chrisunl ,

I think we can get this PR merged in `master`.
Then you can create another branch only for the tests.

**Pull request merged**